### PR TITLE
Improved Image Selection functionality

### LINKED ISSUE #46: [Update] Image Selection feature
We need to score images based on their relevancy. 
We also need to provide multiple images to the end user.

### PR DESCRIPTION
**Pull request by Krechals**
I managed to improve image selection using **Cloud Vision API**.
Images are analysed based on labels, landmarks, text appearance or logos. The Analyser class follows **Strategy Design Pattern** --> their subclasses: **LabelAnalyser**, **LandmarkAnalyser**, **OcrAnalyser**, **LogoAnalyser**.
Using the data extracted from the API, I created a **scoring algorithm** so that relevant images receive high score.

In order to make sure that the app works correctly, multiple unit tests have been added.

### CODE REVIEW

File: src/main/java/com/google/sps/data/TextAnalyser.java [hunk 1]
```diff
@@ -208,7 +208,6 @@ public boolean isInjection() {
       keyWords.addAll(getEvents());
       keyWords.addAll(getEntities());
       keyWords.addAll(getCategories());
-      keyWords.add(getMood());
```
Comment (maraPislar) [inline src/main/java/com/google/sps/data/TextAnalyser.java:211]: I think the mood is still an useful keyword, but you deleted it. What's your opinion?

File: src/main/java/com/google/sps/image/ImageScorer.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  private static final float LANDMARK_SCORE = 0.8f;
+
+  private static final float LOGO_SCORE = 0.5f;
+
+  private static final float OCR_SCORE = 0.1f;
+
+  private static final float LABEL_SCORE = 0.2f;
+
+  private static final double EPS = 1e-5;
+
```
Comment (maraPislar) [inline src/main/java/com/google/sps/image/ImageScorer.java:18]: The new line between these constants seems unusual to me. You can check this as an example maybe https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names
I didn't find any style guide that specifies the mandatory new line between variable declarations, but I can see that you declare all your class variables like this.

File: src/main/java/com/google/sps/image/ImageScorer.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        if (labelElements.contains(keyword)) {
+          score += LABEL_SCORE;
+        }
+      }
+    }
+    return score;
+  }
+
+  /**
+   *  The algorithm that socres images.
```
Comment (maraPislar) [inline src/main/java/com/google/sps/image/ImageScorer.java]: notice: *scores

File: src/main/java/com/google/sps/image/ImageSelection.java [hunk 4]
```diff
@@ -48,47 +60,92 @@ public String generateSearchUrl() {
...
    * This is an endpoint. Call this function to get a relevant image.
    *
-   * @return  URL of the first image scraped from Bing Image Search.
+   * @param     analysationDepth Maximm number of different queries that the web crawler does. 
+   * @return    URL of the first image scraped from Bing Image Search.
    * @exception IOException if Bing doesn't map any image to the keywords.
    */
-  public String getBestImage() throws IOException {
-    List<String> imgSrc = new ArrayList<>();
+  public String getBestImage(int analysationDepth) throws IOException {
```
Comment (maraPislar) [inline src/main/java/com/google/sps/image/ImageSelection.java:100]: This is a really long method, maybe try and split it in multiple methods for readability (if possible)

File: src/main/java/com/google/sps/servlets/InputServlet.java [hunk 5]
```diff
@@ -63,11 +65,11 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     }
 
     TextAnalyser textAnalyser = new TextAnalyser(input_text);
-    Set<String> keywords = textAnalyser.getKeyWords();
-    keywords.add(user_location);
-    ImageSelection imageSelect = new ImageSelection(keywords);
+    Set<String[]> keywordSet = textAnalyser.getSetsOfKeyWords();
```
Comment (maraPislar) [inline src/main/java/com/google/sps/servlets/InputServlet.java:68]: Location of the user must be included, but I will pull request it later today.

File: src/test/java/com/google/sps/LocalImageScorer.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        if (labelElements.contains(keyword)) {
+          score += LABEL_SCORE;
+        }
+      }
+    }
+    return score;
+  }
+
+  /**
+   *  The algorithm that socres images.
```
Comment (maraPislar) [inline src/test/java/com/google/sps/LocalImageScorer.java]: notice: *scores

**Review by maraPislar — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/image/ImageScorer.java (hunk 2 above)
Comment (Krechals) [reply src/main/java/com/google/sps/image/ImageScorer.java:18]: The Google Style Guide says this is optional. I took into account the readability, but I can create blocks of elements that refer to the same topic (label, landmark, ocr, logo scores). https://google.github.io/styleguide/javaguide.html#s4.6.1-vertical-whitespace

File: src/main/java/com/google/sps/data/TextAnalyser.java (hunk 1 above)
Comment (Krechals) [reply src/main/java/com/google/sps/data/TextAnalyser.java:211]: I tested the functionality many times and it does not improve the quality of the keyword set ("happy birthday thrilled" / "happy birthday happy", "job promotion happy"), but it decreases it. By adding the mood, there will also be multiple additional keyword sets:
equation, where N is the number of all keywords).

Comment (Krechals) [reply src/main/java/com/google/sps/data/TextAnalyser.java:211]: There are also cases when the mood improves the quality of keyword set, but we must clearly define that pattern and add the mood just when it's useful.

Comment (maraPislar) [reply src/main/java/com/google/sps/data/TextAnalyser.java:211]: Okay, I see. I will figure out a pattern.

File: src/test/java/com/google/sps/LocalImageScorer.java (hunk 6 above)
Comment (Krechals) [reply src/test/java/com/google/sps/LocalImageScorer.java]: Done

File: src/main/java/com/google/sps/image/ImageSelection.java (hunk 4 above)
Comment (Krechals) [reply src/main/java/com/google/sps/image/ImageSelection.java:100]: Done.

File: src/main/java/com/google/sps/image/ImageScorer.java (hunk 3 above)
Comment (Krechals) [reply src/main/java/com/google/sps/image/ImageScorer.java]: Done.

**Review by maraPislar — APPROVED** [review]

**Review by AndreMB0 — APPROVED** [review]
LGTM.
remember next time to submit small PRs :)

**Pull request merged**